src/Plugin/EntityReferenceSelection/OgSelection.php

```diff
@@ -81,37 +155,45 @@ protected function buildEntityQuery($match = NULL, $match_operator = 'CONTAINS')
       $query->condition($bundle_key, $bundles, 'IN');
     }
 
-    $user_groups = $this->getUserGroups();
-    if (!$user_groups) {
+    // Get the identifier key of the entity.
+    $identifier_key = $definition->getKey('id');
+
+    if ($this->currentUser->isAnonymous()) {
+      // @todo Check if anonymous users should have access to any referenced
+      // groups? What about groups that allow anonymous posts?
+      return $query->condition($identifier_key, -1);
+    }
+
+    if ($this->currentUser->hasPermission('administer organic groups')) {
+      // User can see all the groups.
       return $query;
     }
 
-    $identifier_key = $definition->getKey('id');
+    if (empty($this->configuration['entity'])) {
+      // @todo Find out why we have this scenario.
+      return $query;
+    }
+
+    /** @var \Drupal\Core\Entity\EntityInterface $entity */
+    $entity = $this->configuration['entity'];
+    $bundle = $entity->bundle();
 
     $ids = [];
-    if (!empty($this->getConfiguration()['field_mode']) && $this->getConfiguration()['field_mode'] === 'admin') {
-      // Don't include the groups, the user doesn't have create permission.
-      foreach ($user_groups as $group) {
+    foreach ($this->getUserGroups() as $group) {
+      // Check user has "create" permission on this entity.
+      $permission = "create $bundle content";
```

Comment:
@zero2one, we should use the permission layer instead of using string patterns:
```suggestion
      // Check if the current user can perform the "create" operation on this entity.
      if ($this->ogAccess->userAccessGroupContentEntityOperation('create', $group, $entity, $this->currentUser)) {
```

Reply:
Then we need to check if we have a new or existing entity ($entity->isNew()) and do something like this:

* Entity is new => check "create" operation for that group.
* Entity is existing and linked to the group being checked: check "update" operation for that group.
* Entity is existing and not linked to the group being checked : check "create" operation for that group.